astropy/convolution/src/boundary_none.c

```diff
@@ -0,0 +1,513 @@
+
+// 3D
+void convolve3d_boundary_none_c(DTYPE * const result,
+        const DTYPE * const f, const size_t nx, const size_t ny, const size_t nz,
+        const DTYPE * const g, const size_t nkx, const size_t nky, const size_t nkz,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+inline __attribute__((always_inline)) void convolve3d_boundary_none(DTYPE * const result,
+        const DTYPE * const f, const size_t nx, const size_t ny, const size_t nz,
+        const DTYPE * const g, const size_t nkx, const size_t nky, const size_t nkz,
+        const bool nan_interpolate,
+        const unsigned n_threads);
+
+void convolveNd_boundary_none_c(DTYPE * const result,
+        const DTYPE * const f,
+        const unsigned n_dim,
+        const size_t * const image_shape,
+        const DTYPE * const g,
+        const size_t * const kernel_shape,
+        const bool nan_interpolate,
+        const unsigned n_threads)
+{
+#ifdef NDEBUG
+    if (!result || !f || !g || !image_shape || !kernel_shape)
+        return;
+#else
+    assert(result);
+    assert(f);
+    assert(g);
+    assert(image_shape);
+    assert(kernel_shape);
+#endif
+
+    if (n_dim == 1)
+        convolve1d_boundary_none_c(result, f,
+                image_shape[0],
+                g, kernel_shape[0],
+                nan_interpolate, n_threads);
+    else if (n_dim == 2)
+        convolve2d_boundary_none_c(result, f,
+                image_shape[0], image_shape[1],
+                g, kernel_shape[0], kernel_shape[1],
+                nan_interpolate, n_threads);
+    else if (n_dim == 3)
+        convolve3d_boundary_none_c(result, f,
+                        image_shape[0], image_shape[1], image_shape[2],
+                        g, kernel_shape[0], kernel_shape[1], kernel_shape[2],
+                        nan_interpolate, n_threads);
+    else
+        assert(0); // Unimplemented: n_dim > 3
+}
+
+/*-------------------------PERFORMANCE NOTES--------------------------------
+ * The function wrappers below are designed to take advantage of the following:
+ * The preprocessor will inline convolve<N>d_boundary_none(), effectively
+ * expanding the two logical branches, replacing nan_interpolate
+ * for their literal equivalents. The corresponding conditionals
+ * within these functions will then be optimized away, this
+ * being the goal - removing the unnecessary conditionals from
+ * the loops without duplicating code.
+ *--------------------------------------------------------------------------
+ */
+
+void convolve1d_boundary_none_c(DTYPE * const result,
+        const DTYPE * const f, const size_t nx,
+        const DTYPE * const g, const size_t nkx,
+        const bool nan_interpolate,
+        const unsigned n_threads)
+{
+#ifdef NDEBUG
+    if (!result || !f || !g)
+        return;
+#else
+    assert(result);
+    assert(f);
+    assert(g);
+#endif
+
+    if (nan_interpolate)
+        convolve1d_boundary_none(result, f, nx, g, nkx, true, n_threads);
```

Comment:
Looks like this could be rewritten as a single statement: `convolve1d_boundary_none(result, f, nx, g, nkx, nan_interpolate, n_threads);` rather than a `if` condition

Reply:
Or is the intention that the compiler propagates the constant? (Does that actually happen in practice without something like `-flto`?)

Reply:
No, this needs to stay as it is. It is one of the primary optimizations. `convolve1d_boundary_none` is inlined (always) and the point is that the preprocessor will insert the boolean literals into the inline code when it expands the function, twice, once for `true`, the other for `false'.  These literals then allow the compiler to optimize out the expensive conditionals involved for the NaN interpolation logic for the `false` branch, by shortcircuiting that logic. Basically, it's a neat trick to keep the code as a single function without having to write two separate functions, one with the required conditional logic and one without - the compiler writes this for us. See the [comment within the code](https://github.com/jamienoss/astropy/blob/2020741719394da9321d0ca602dc20d4a1ab8217/astropy/convolution/src/boundary_none.c#L106-L115):
 ```c
/*-------------------------PERFORMANCE NOTES--------------------------------
 * The function wrappers below are designed to take advantage of the following:
 * The preprocessor will inline convolve<N>d_boundary_none(), effectively
 * expanding the two logical branches, replacing nan_interpolate
 * for their literal equivalents. The corresponding conditionals
 * within these functions will then be optimized away, this
 * being the goal - removing the unnecessary conditionals from
 * the loops without duplicating code.
 *--------------------------------------------------------------------------
 */
`

Reply:
I am curious - does this actually work in practice -- i.e., if you add the `-E` flag to the `gcc` and inspect the post pre-processor C source, do you see both the "true" and the "false"  values inlined separately? What's the difference in speed between the existing implementation and naively calling (like I suggested): 
`convolve1d_boundary_none(result, f, nx, g, nkx, nan_interpolate, n_threads); `

I would think that since `nan_interpolate` is a runtime constant, the branch predictor will always get that branch correct after the first few attempts. Then, the presence of the additional `isnan()` code does not matter in the `nan_interpolate=False` case (unless you are thinking about the instruction cache, but 3 instructions probably isn't going to hurt the instruction cache)

Reply:
> I am curious - does this actually work in practice 

Most definitely, it is a simple and commonly used pattern.

> I am curious - does this actually work in practice -- i.e., if you add the -E flag to the gcc and inspect the post pre-processor C source, do you see both the "true" and the "false" values inlined separately?

Yes, (depending on exactly what you mean) if you didn't see this then it wasn't inlined. The code has to exist twice, it has to get expanded twice. It's the preprocessor that does this, and they're not intelligent by design, they're lexical translators. Think of it as a macro (and then as soon as you have, shudder and never do it again). Even if the literals aren't replaced in their respective conditional lines of code, they have to persist - e.g.
```c
if (nan_interpolate)
{
    tempBool = true;
    ...
    // expanded function
    ...
    if (tempBool)
    {
        if (!isnana(val))
        ...
    }
    ...
}
else
{     
    tempBool = false;
    ...
    // expanded function
    ...
    if (tempBool)
    {
        if (!isnana(val))
        ...
    }
    ...
}
```
Whether it's the code above or below, it doesn't matter:
```c
if (nan_interpolate)
{ 
    ...
    if (true)
    {
        if (!isnana(val))
        ...
    }
    ...
}
else
{     
...
}
```

>I would think that since nan_interpolate is a runtime constant, the branch predictor will always get that branch correct after the first few attempts.

Maybe, maybe not, but anyhow that's not really the point. Procs don't vectorize your loops for you, runtime constant or not. Compilers do though, but (at least) only for condition free loops.

>What's the difference in speed between the existing implementation and naively calling

Firstly, the naive approace is how it existed before. I think, there's a chance that you might be focusing on the wrong branch of the condition. There is next to no perf boost for `nan_interpolate == true`.  The fact that these conditionals get optimized out is, in the big picture irrelevant, they're just a byproduct of the pattern. The perf boost comes from the fact that `nan_interpolate` is on by default. So each pix is being checked for NaNs even though there might not be any. There's a quick check in `convolve.py::convolve()` ([here](https://github.com/jamienoss/astropy/blob/2020741719394da9321d0ca602dc20d4a1ab8217/astropy/convolution/convolve.py#L265-L269)) to see if any NaNs do exist, if they don't it turns the interpolation off. 
```python
    nan_interpolate = (nan_treatment == 'interpolate') and np.isnan(array_internal.sum())
```
Then comes the "sigh" and "ugh", I'm gonna have to write two funcs, one without the interpolation code, "yuk!". Nope! Make it inline (always), and just wrap it with boolean literals - job done.
For the array and kernel sizes I used for my bench, the perf increase was ~15% (if I rememebr correctly - it's in the notes at the top of the PR).